File: Python_code/curve_draw.py
import numpy as np
import matplotlib.pyplot as plt

# Define the closed curve in 3D
t = np.linspace(0, 2*np.pi, 100)
x = np.cos(t)
y = np.sin(t)
z = np.cos(2*t)

# Find the highest and lowest points of the curve
z_max_index = np.argmax(z)
z_min_index = np.argmin(z)

# Define the sphere
phi, theta = np.linspace(0, np.pi, 20), np.linspace(0, 2*np.pi, 40)
phi, theta = np.meshgrid(phi, theta)
x_sphere = 0.2 * np.sin(phi) * np.cos(theta)
y_sphere = 0.2 * np.sin(phi) * np.sin(theta)
z_sphere = 0.2 * np.cos(phi)

# Generate the sweeping surface
fig = plt.figure()
ax = fig.add_subplot(111, projection='3d')
ax.plot_surface(x_sphere, y_sphere, z_sphere)

# Move the sphere along the closed curve
for i in range(len(t)):
    # Translate the sphere to a new position
    x_sphere_new = x_sphere + x[i]
    y_sphere_new = y_sphere + y[i]
    z_sphere_new = z_sphere + z[i]

    # Update the surface
    ax.clear()
    ax.plot_surface(x_sphere_new, y_sphere_new, z_sphere_new)

    # Draw the closed curve
    ax.plot(x, y, z, 'r', linewidth=3)

    # Draw the highest and lowest points in blue
    if i == z_max_index:
        ax.plot(x[i], y[i], z[i], 'bo', markersize=10)
    elif i == z_min_index:
        ax.plot(x[i], y[i], z[i], 'bo', markersize=10)
    else:
        ax.plot(x[i], y[i], z[i], 'ro', markersize=10)

    # Update the plot
    ax.set_xlim([-1.5, 1.5])
    ax.set_ylim([-1.5, 1.5])
    ax.set_zlim([-1.5, 1.5])
    plt.draw()
    plt.pause(0.1)

File: matlab_code/curve_draw.m
% Define the closed curve in 3D
t = linspace(0,2*pi,100);
x = cos(t);
y = sin(t);
z = cos(2*t);

% Find the minimum and maximum values of z
z_min = min(z);
z_min_index = find(z == z_min);
z_max = max(z);
z_max_index = find(z == z_max);

% Define the sphere
diameter = 0.2;
[X,Y,Z] = sphere(30);
X = X * diameter/2;
Y = Y * diameter/2;
Z = Z * diameter/2;

% Generate the sweeping surface
h = surf(X,Y,Z);
axis equal
hold on

% Move the sphere along the closed curve
for i = 1:length(t)
    % Translate the sphere to a new position
    X_new = X + x(i);
    Y_new = Y + y(i);
    Z_new = Z + z(i);
    
    % Update the surface
    set(h,'XData',X_new,'YData',Y_new,'ZData',Z_new);
    
    % Draw the closed curve
    plot3(x,y,z,'r','LineWidth',3);
    
    % Highlight the minimum and maximum points of z
    plot3(x(z_min_index),y(z_min_index),z(z_min_index),'bo','MarkerSize',10,'MarkerFaceColor','b');
    plot3(x(z_max_index),y(z_max_index),z(z_max_index),'bo','MarkerSize',10,'MarkerFaceColor','b');
    
    % Draw the sphere
    plot3(x(i),y(i),z(i),'ro','MarkerSize',10,'MarkerFaceColor','r');
    
    % Update the plot
    pause(0.1);
    drawnow;
end